src/basic/virt.c

```diff
@@ -472,8 +463,18 @@ int detect_vm(void) {
                 return r;
         if (r == VIRTUALIZATION_VM_OTHER)
                 other = true;
-        else if (r != VIRTUALIZATION_NONE)
-                goto finish;
+        else if (r != VIRTUALIZATION_NONE) {
```

Comment:
the assumption here is that r is VIRTUALIZATION_XEN. Maybe make this explicit via an assert(), i.e.

```c
…
else if (r != VIRTUALIZATION_NONE) {
        assert(r == VIRTUALIZATION_XEN);
        …
```

or maybe:

```
else if (r == VIRTUALIZATION_XEN) {
        …
else if (r != VIRTUALIZATION_NONE)
        assert_not_reached();
```

or so

Reply:
Makes sense to make it explicit. I notice that two checks:

detect_vm_uml
detect_vm_xen

Have no code path which returns VIRTUALIZATION_VM_OTHER. So I have also taken the opportunity to tidy the conditionals to not check this (for example !none and !xen should remain an assertion failure).